lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java

```diff
@@ -0,0 +1,217 @@
+
+  private final IdentifierFactory idFactory;
+  private final NameResolver nameResolver;
+  private final Transport transport;
+  private final EventHandler<TransportEvent> recvHandler;
+  private final ConcurrentMap<String, NSConnectionFactory> connectionFactoryMap;
+  private final ConcurrentMap<String, Boolean> isStreamingCodecMap;
+  private Identifier myId;
+  private final Codec<DefaultNSMessage> nsCodec;
+  private final LinkListener<DefaultNSMessage> nsLinkListener;
+  private final EStage<Tuple<Identifier, InetSocketAddress>> nameServiceRegisteringStage;
+  private final EStage<Identifier> nameServiceUnregisteringStage;
+
+  @Inject
+  private DefaultNetworkServiceImpl(
+      @Parameter(NetworkServiceParameters.NetworkServiceIdentifierFactory.class) final IdentifierFactory idFactory,
+      @Parameter(NetworkServiceParameters.NetworkServicePort.class) final int nsPort,
+      final TransportFactory transportFactory,
+      final NameResolver nameResolver) {
+
+    this.idFactory = idFactory;
+    this.connectionFactoryMap = new ConcurrentHashMap<>();
+    this.isStreamingCodecMap = new ConcurrentHashMap<>();
+    this.nsCodec = new DefaultNSMessageCodec(idFactory, connectionFactoryMap, isStreamingCodecMap);
+    this.nsLinkListener = new NetworkServiceLinkListener(connectionFactoryMap);
+    this.recvHandler = new NetworkServiceReceiveHandler(connectionFactoryMap, nsCodec);
+    this.nameResolver = nameResolver;
+    this.transport = transportFactory.newInstance(nsPort, recvHandler, recvHandler, new DefaultNSExceptionHandler());
+
+    this.nameServiceRegisteringStage = new SingleThreadStage<>(
+        "NameServiceRegisterer", new EventHandler<Tuple<Identifier, InetSocketAddress>>() {
+      @Override
+      public void onNext(final Tuple<Identifier, InetSocketAddress> tuple) {
+        try {
+          nameResolver.register(tuple.getKey(), tuple.getValue());
+          LOG.log(Level.FINEST, "Registered {0} with nameservice", tuple.getKey());
+        } catch (final Exception ex) {
+          final String msg = "Unable to register " + tuple.getKey() + " with name service";
+          LOG.log(Level.WARNING, msg, ex);
+          throw new RuntimeException(msg, ex);
+        }
+      }
+    }, 5);
+
+    this.nameServiceUnregisteringStage = new SingleThreadStage<>(
+        "NameServiceRegisterer", new EventHandler<Identifier>() {
+      @Override
+      public void onNext(final Identifier id) {
+        try {
+          nameResolver.unregister(id);
+          LOG.log(Level.FINEST, "Unregistered {0} with nameservice", id);
+        } catch (final Exception ex) {
+          final String msg = "Unable to unregister " + id + " with name service";
+          LOG.log(Level.WARNING, msg, ex);
+          throw new RuntimeException(msg, ex);
+        }
+      }
+    }, 5);
+  }
+
+  @Override
+  public synchronized <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
```

Comment:
`getConnectionFactory` is just to read map. I don't think we have to `synchronize` this method.